api/src/main/java/org/pmiops/workbench/api/DataSetController.java

```diff
@@ -300,101 +299,88 @@ private DataSetRequest generateDataSetRequestFromPreviewRequest(
-                          .findFirst()
-                          .get();
-                  if (fields.getType() == LegacySQLTypeName.TIMESTAMP) {
-                    List<String> queryValues = new ArrayList<String>();
-                    DateFormat dateFormat = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss");
-                    previewValue
-                        .getQueryValue()
-                        .forEach(
-                            value -> {
-                              try {
-                                Double fieldValue = Double.parseDouble(value);
-                                queryValues.add(
-                                    dateFormat.format(new Date(fieldValue.longValue())));
-                              } catch (NumberFormatException ex) {
-                                queryValues.add("");
-                              }
-                            });
-                    previewValue.setQueryValue(queryValues);
-                  }
-                });
-        break;
-      } catch (Exception ex) {
-        if ((ex.getCause() != null
-            && ex.getCause().getMessage() != null
-            && ex.getCause().getMessage().contains("Read timed out"))) {
-          rowsRequested = (rowsRequested / 2);
-          if (rowsRequested == 0) {
-            throw new GatewayTimeoutException(
-                "Timeout while querying the CDR to pull preview information.");
-          }
-          retry++;
-        } else {
-          throw ex;
-        }
+    TableResult queryResponse;
+    try {
+      String query = originalQuery.concat(" LIMIT " + NO_OF_PREVIEW_ROWS);
+
+      queryJobConfiguration = queryJobConfiguration.toBuilder().setQuery(query).build();
+
+      /* Google appengine has a 60 second timeout, we want to make sure this endpoint completes
+       * before that limit is exceeded, or we get a 500 error with the following type:
+       * com.google.apphosting.runtime.HardDeadlineExceededError
+       * See https://cloud.google.com/appengine/articles/deadlineexceedederrors for details
+       */
+      queryResponse =
+          bigQueryService.executeQuery(
+              bigQueryService.filterBigQueryConfig(queryJobConfiguration),
+              APP_ENGINE_HARD_TIMEOUT_MSEC_MINUS_FIVE_SEC);
+    } catch (Exception ex) {
+      if ((ex.getCause() != null
+          && ex.getCause().getMessage() != null
+          && ex.getCause().getMessage().contains("Read timed out"))) {
+        throw new GatewayTimeoutException(
+            "Timeout while querying the CDR to pull preview information.");
+      } else {
+        throw ex;
       }
-    } while (retry < PREVIEW_RETRY_LIMIT);
-
-    if (retry == PREVIEW_RETRY_LIMIT) {
-      throw new GatewayTimeoutException(
-          "Timeout while querying the CDR to pull preview information.");
     }
+    queryResponse
+        .getSchema()
+        .getFields()
+        .forEach(
```

Comment:
nit: Instead of foreach, you can map/collect into a list, and then do an addAll onto `valuePreviewList`. The main difference for me is that readers generally expect `forEach` loops to have side effects, and thus be less functional than the stream style.

Reply:
Will change.